### PR TITLE
fix(coverity): use xstrndup() instead of vim_strsave()

### PR DESCRIPTION
**Pull request by zeertzjq**

### CODE REVIEW

**Comment by zeertzjq** [comment]
There are several similar occurrences:
```
rc/nvim/lua/executor.c
1652:              vim_strsave((char_u *)v.data.string.data));

src/nvim/api/autocmd.c
463:      cb.data.funcref = vim_strsave((char_u *)callback->data.string.data);
477:      aucmd.callable.cmd = vim_strsave((char_u *)command->data.string.data);
817:    pattern = vim_strsave((char_u *)opts->pattern.data.string.data);

src/nvim/api/vim.c
2525:  char_u *cmdline = vim_strsave((char_u *)str.data);
```

I guess there needs to be a helper function to save a `String`?

**Review by justinmk — APPROVED** [review]

**Comment by zeertzjq** [comment]
Does this really fix the coverity warning?

**Comment by zeertzjq** [comment]
I added a `string_to_cstr()` function to avoid repeating the variable name twice.

File: src/nvim/api/private/helpers.c [hunk 1]
```diff
@@ -488,6 +488,15 @@ String cstr_to_string(const char *str)
   };
 }
 
+/// Copies a String to an allocated, NUL-terminated C string.
+///
+/// @param string the String to copy
+/// @return the resulting C string
+char *string_to_cstr(String string)
```
Comment (justinmk) [inline src/nvim/api/private/helpers.c:495]: nit: looks like `str` is more common in this file
```suggestion
char *string_to_cstr(String str)
```

**Review by justinmk — APPROVED** [review]

**Comment by zeertzjq** [comment]
There are actually many more cases where an API function passes the `data` of a `String` to a function assuming a NUL-terminated string, for example:
```c
Dictionary nvim_get_hl_by_name(String name, Boolean rgb, Error *err)
  FUNC_API_SINCE(3)
{
  Dictionary result = ARRAY_DICT_INIT;
  int id = syn_name2id(name.data);

  if (id == 0) {
    api_set_error(err, kErrorTypeException, "Invalid highlight name: %s",
                  name.data);
    return result;
  }
  result = nvim_get_hl_by_id(id, rgb, err);
  return result;
}
```
But the `vim_strsave()` case is the easiest to deal with.

**Comment by zeertzjq** [comment]
Related: #17529

**Pull request merged**